src/kernels/MIOpenSparseSoftmaxCrossEntropyWithLogits.cpp

```diff
@@ -0,0 +1,280 @@
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+#ifndef MIOPEN_DONT_USE_HIP_RUNTIME_HEADERS
+#include <hip/hip_fp16.h>
+#include <hip/hip_runtime.h>
+#endif
+
+#ifndef T_TYPE
+#define T_TYPE int
+#endif
+
+#ifndef DTYPE_LOWEST
+#define DTYPE_LOWEST 0
+#endif
+
+#include "float_types.h"
+#include "tensor_view.hpp"
+
+template <typename T, typename Ta>
+__device__ void sparseSoftmaxCrossEntropyWithLogitsForward(const T* input,
+                                                           const Ta* target,
+                                                           T* output,
+                                                           T* backprop,
+                                                           uint64_t num_class,
+                                                           tensor_view_t<2> input_tv,
+                                                           tensor_view_t<1> target_tv,
+                                                           tensor_view_t<1> output_tv,
+                                                           tensor_view_t<2> backprop_tv)
+{
+    uint64_t gid = blockIdx.x;
+    uint64_t lid = threadIdx.x;
+
+    __shared__ FLOAT_ACCUM lmax[LOCAL_SIZE], lsum[LOCAL_SIZE];
+    lmax[lid] = DTYPE_LOWEST;
+    lsum[lid] = 0.0f;
+    __shared__ uint64_t label;
+
+    if(lid == 0)
+    {
+        label = static_cast<uint64_t>(target[target_tv.get_tensor_view_idx({gid})]);
+    }
+
+    for(uint64_t i = lid; i < num_class; i += LOCAL_SIZE)
+    {
+        FLOAT_ACCUM val = CVT_FLOAT2ACCUM(input[input_tv.get_tensor_view_idx({gid, i})]);
+        lmax[lid]       = max(lmax[lid], val);
+    }
+    __syncthreads();
+
+    for(uint64_t i = LOCAL_SIZE >> 1; i > 0; i >>= 1)
+    {
+        if(lid < i)
+        {
+            lmax[lid] = max(lmax[lid], lmax[lid + i]);
+        }
+        __syncthreads();
+    }
+
+    for(uint64_t i = lid; i < num_class; i += LOCAL_SIZE)
+    {
+        FLOAT_ACCUM val = CVT_FLOAT2ACCUM(input[input_tv.get_tensor_view_idx({gid, i})]);
+        lsum[lid] += exp(val - lmax[0]);
+    }
+    __syncthreads();
+
+    for(uint64_t i = LOCAL_SIZE >> 1; i > 0; i >>= 1)
+    {
+        if(lid < i)
+        {
+            lsum[lid] += lsum[lid + i];
+        }
+        __syncthreads();
+    }
```

Comment:
This is reduction_sum at Block level. Replace it with `block_reduce` template function in `src/kernels/block_reduce.hpp`.

See `ReduceSumFLOATACCUM` at `MIOpenReduceSum.cpp` for an example.

Reply:
resolved